allennlp/data/token_indexers/pretrained_transformer_indexer.py

```diff
@@ -83,14 +87,39 @@ def tokens_to_indices(self, tokens: List[Token], vocabulary: Vocabulary) -> Inde
                     f" for the following token: {token.text}"
                 )
 
+            if type_ids is not None and getattr(token, "type_id", None) is not None:
+                type_ids.append(token.type_id)
+            else:
+                type_ids = None
+
         # The mask has 1 for real tokens and 0 for padding tokens. Only real tokens are attended to.
         mask = [1] * len(indices)
 
-        return {"token_ids": indices, "mask": mask}
+        result = {"token_ids": indices, "mask": mask}
+        if type_ids is not None:
+            result["type_ids"] = type_ids
+        return result
 
     @overrides
     def get_empty_token_list(self) -> IndexedTokenList:
-        return {"token_ids": [], "mask": []}
+        return {"token_ids": [], "mask": [], "type_ids": []}
+
+    @overrides
+    def as_padded_tensor_dict(
+        self, tokens: IndexedTokenList, padding_lengths: Dict[str, int],
+    ) -> Dict[str, torch.Tensor]:
+        # Different transformers use different padding values for tokens, but for mask and type id, the padding
+        # value is always 0.
+        return {
+            key: torch.LongTensor(
+                pad_sequence_to_length(
+                    val,
+                    padding_lengths[key],
+                    default_value=lambda: 0 if key in {"mask", "type_ids"} else self._padding_value,
```

Comment:
Wait so were we broken for the mask case?

Reply:
Yeah, could you add a test for this because it's pretty critical.

Reply:
RoBERTa pads with 1s. We probably have other places too where we assume the padding value is always 0.

Reply:
89cd484d17affbeda989db2939da57c21db3471e